### PR TITLE
feat(cache): add cache:renew function and mlcache refactorings

### PR DESCRIPTION
**Pull request by bungle**
### Summary

1. removes fallbacks needed for older OpenResty support
2. uses SHM native TTLs only instead, and stores just the value in SHM instead of storing `<type>:<created_at>:<ttl>:<value>`.
3. implements `:renew` in mlcache
4. implements `:renew` in kong cache
5. changes kong.vault to use `:renew` function instead of calling `:peek` etc. for better concurrency control

KAG-1698

### CODE REVIEW

**Review by hanshuebner — CHANGES_REQUESTED** [review]
Looks good - Following up to our discussion yesterday, I think it would be better to replace the use of `rotate` by `renew`.

**Comment by chronolaw** [comment]
Could we still keep `kong.cache.marshall` as an interlayer? Though it is just a alias of `string.buffer.encode`.

**Comment by hanshuebner** [comment]
> Could we still keep `kong.cache.marshall` as an interlayer? Though it is just a alias of `string.buffer.encode`.

Why do you want to keep it?

**Comment by bungle** [comment]
> Looks good - Following up to our discussion yesterday, I think it would be better to replace the use of `rotate` by `renew`.

@hanshuebner, I will make a change and ping you back.

**Review by hanshuebner — APPROVED** [review]

File: kong/resty/mlcache/init.lua [hunk 1]
```diff
@@ -888,6 +783,186 @@ function _M:get(key, opts, cb, ...)
...
+
+    v, shmerr = self.dict:get_stale(namespaced_key)
+    if v ~= nil then
+        is_hit = true
+
+    else
+        if shmerr then
+            -- shmerr can be 'flags' upon successful get_stale() calls, so we
+            -- also check v == nil
+            return nil, "could not read from lua_shared_dict: " .. shmerr
```
Comment (chobits) [inline kong/resty/mlcache/init.lua:850]: hi @bungle do we need to unlock `lock` object here before it returns nil?

Comment (chobits) [reply kong/resty/mlcache/init.lua:850]: Seems no bug here, from the lua-resty-lock doc:


> If the [unlock()](https://github.com/openresty/lua-resty-lock#unlock) method is never called after this method call, the lock will get released when
> 
> the current resty.lock object instance is collected automatically by the Lua GC.
the exptime for the lock entry is reached.



but I think we could `call unlock_and_ret()` to align with the `get()` api's action.

**Pull request merged**